rust/diesel/src/order.rs

```diff
@@ -0,0 +1,173 @@
+use bigdecimal::BigDecimal;
+use diesel;
+use diesel::pg::PgConnection;
+use diesel::prelude::*;
+use serde::{Deserialize, Serialize};
+use std::ops::{Add, Mul};
+
+use crate::product::Product;
+use crate::schema::order_lines;
+use crate::schema::orders;
+use num_traits::identities::Zero;
+
+#[derive(AsChangeset, Serialize, Deserialize, Queryable, Insertable)]
+#[serde(rename_all = "camelCase")]
+pub struct Order {
+    pub order_id: i32,
+    pub order_total: BigDecimal,
+    pub user_id: i32,
+}
+
+#[table_name = "orders"]
+#[derive(AsChangeset, Serialize, Deserialize, Queryable, Insertable)]
+#[serde(rename_all = "camelCase")]
+struct NewOrder {
+    pub order_total: BigDecimal,
+    pub user_id: i32,
+}
+
+#[table_name = "order_lines"]
+#[derive(AsChangeset, Serialize, Deserialize, Queryable, Insertable)]
+#[serde(rename_all = "camelCase")]
+pub struct NewOrderLine {
+    pub order_id: i32,
+    pub product_id: i32,
+    pub units: i16,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct NewUserOrder {
+    pub user_id: i32,
+    pub products: Vec<ProductOrder>,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ProductOrder {
+    pub product_id: i32,
+    pub units: i16,
+}
+
+impl Order {
+    pub fn create(order: NewUserOrder, connection: &PgConnection) -> Option<NewUserOrder> {
+        let product_units: Vec<(Product, i16)> = order
+            .products
+            .iter()
+            .flat_map(|product_ref| {
+                Product::read(product_ref.product_id, connection)
+                    .map(|product| (product, product_ref.units))
+            })
+            .collect();
+
+        let order_total = product_units
+            .to_vec()
+            .into_iter()
+            .fold(BigDecimal::from(0.0), |acc, (product, qty)| {
+                acc.add(product.price.mul(BigDecimal::from(qty)))
+            });
+
+        if order_total.is_zero() {
+            return None;
+        }
+
+        let row = NewOrder {
+            user_id: order.user_id,
+            order_total,
+        };
+
+        let insert_result = connection.build_transaction().read_write().run(|| {
+            let inserted_order: Order = diesel::insert_into(orders::table)
+                .values(&row)
+                .get_result(connection)
+                .unwrap();
+
+            let order_lines: Vec<NewOrderLine> = product_units
+                .to_vec()
+                .into_iter()
```

Comment:
I left `clone().into_iter()` as it copies array and moves value into lambda, more efficient than borrowing via `iter()` and cloning each value

Reply:
`order`, `inserted_order.order_id`, and `product.product_id` are all either `i32` or `i16`, so they are trivially copy-able. So you _can_ use `.iter()` without having to clone anything, and it will be just as fast as moving the values, and faster than cloning a vector with heap memory.